### PR TITLE
CI(openwrt): refactor build and test process

### PR DESCRIPTION
**Pull request by 1715173329**
This refactored the whole build & test process.

Cleaned up the Makefile, and switched to cmake build.

For CI, switched to OpenWrt's official build process, and build ipk for popular architectures.
Also implemented runtime test for armv7 / aarch64 and i386 / amd64.

### CODE REVIEW

**Comment by hamishcoleman** [comment]
I think that there are some good things here, but I am concerned about accepting all of the change.  It would be easier if it were broken down into smaller - functionally separate - changes that we could review one at a time.

For a start, you have switched the build to use the deprecated cmake system - the Makefiles are the preferred build process, so we dont want to increase the usage of the cmake.

**Comment by 1715173329** [comment]
> I think that there are some good things here, but I am concerned about accepting all of the change. It would be easier if it were broken down into smaller - functionally separate - changes that we could review one at a time.

Okay I will do it today.

> For a start, you have switched the build to use the deprecated cmake system - the Makefiles  #are the preferred build process, so we dont want to increase the usage of the cmake.

Cmake is the modern build system, I have no idea why deprecated it. Cmake is simpler than legacy autotools, more human readable, and more friendly to crossing-compile. It also compiles faster.
But sure, I can switch to autotools back ;)

**Comment by 1715173329** [comment]
Hi @hamishcoleman, is there any way to build test tools without actually executing them? It's impossible to execute them on the host as we are doing crossing-compile.

```
make[4]: Leaving directory '/home/buildbot/immortalwrt/build_dir/target-aarch64_cortex-a53_musl/n2n-ba8855fa71ca864e180876e0a10088a73b3ce1fa/tools'
scripts/test_harness.sh tests/tests_units.list
./tools/tests-auth >tests/tests-auth.out
/lib/ld-musl-aarch64.so.1: No such file or directory
Makefile:211: recipe for target 'test.units' failed
make[3]: *** [test.units] Error 255
```

**Comment by hamishcoleman** [comment]
> Hi @hamishcoleman, is there any way to build test tools without actually executing them?

The default make target will build all the binaries - including the test tools

**Comment by hamishcoleman** [comment]
@1715173329 Can you point me to the documentation that you refer to for the openwrt official build process?

**Comment by 1715173329** [comment]
> @1715173329 Can you point me to the documentation that you refer to for the openwrt official build process?

https://github.com/openwrt/gh-action-sdk
idk if it can be called as a "documentation ", but no more details available. It's pretty simple I think.

And this is how upstream uses it:
https://github.com/openwrt/packages/tree/master/.github/workflows

**Comment by hamishcoleman** [comment]
Thank you for that pointer.

How are you going with splitting this PR into smaller ones so that each functional change can be separately addressed?

**Comment by 1715173329** [comment]
> The default make target will build all the binaries - including the test tools

Thanks. I found them located in tools, which is different from CMake that puts binaries into root.

> How are you going with splitting this PR into smaller ones so that each functional change can be separately addressed?

All of these changes are around with OpenWrt, I have split them into different commits already.
The latter commit needs the former one to work, I have no idea how to split this PR into even "smaller" ones.

**Comment by hamishcoleman** [comment]
> > The default make target will build all the binaries - including the test tools
> 
> Thanks. I found them located in tools, which is different from CMake that puts binaries into root.

Yes, the CMake files are not well maintained and have not always placed things in the right place.

> > How are you going with splitting this PR into smaller ones so that each functional change can be separately addressed?
> 
> All of these changes are around with OpenWrt, I have split them into different commits already. The latter commit needs the former one to work, I have no idea how to split this PR into even "smaller" ones.

While these are all related to OpenWRT, you are making several different changes all at once.  Some of which we would want to discuss.  If you cannot split this PR then we need to consider it all or nothing, which may not be what you were trying to achieve.  In this case, let me add some review notes.

File: .github/openwrt/Dockerfile
Comment (hamishcoleman) [inline .github/openwrt/Dockerfile:6]: We have not previously used Docker in any of the build processes.  Can you explain exactly what this is going to be used for?  Can we avoid building a Docker image?  Is there somewhere else upstream that has this image available?

File: .github/openwrt/entrypoint.sh
Comment (hamishcoleman) [inline .github/openwrt/entrypoint.sh:10]: This looks like it is hand unpacking a package - is there not a tool that is provided to handle ipk files?

File: .github/openwrt/entrypoint.sh
Comment (hamishcoleman) [inline .github/openwrt/entrypoint.sh:14]: Can you explain what a "variant" and what this "release" is?

File: .github/openwrt/entrypoint.sh
Comment (hamishcoleman) [inline .github/openwrt/entrypoint.sh:26]: Since this commit has included a test.sh, it makes me think that this entrypoint.sh has been copied from a different project - if so, can you document what that project is.  Also, is it possible to share the code with the other project, instead of copying it into this one?

File: .github/openwrt/test.sh
Comment (hamishcoleman) [inline .github/openwrt/test.sh]: This doesnt look like a very useful test.  It would probably make more sense to check the exit code of running the binary.

Note that this also relies on binary names that are non standard within this project (see more comment on this below)

File: .github/openwrt/test.sh
Comment (hamishcoleman) [inline .github/openwrt/test.sh:12]: The functionality contained in this file looks like it might be useful outside the context of a CI build.  Have you considered putting it in the scripts/ subdirectory - or improving one of the scripts already existing for running the tests so it can be used?

File: .github/workflows/openwrt.yml
Comment (hamishcoleman) [inline .github/workflows/openwrt.yml]: Because this file is basically completely rewritten, it is hard to provide a valid review for it.  If the file was incrementally improved to provide the feature set that you are trying to achieve, it would be easier to review.  Failing that, it might be simpler if there were two different files - the file you are rewriting is not broken could exist side-by-side with your replacement.  This would allow a comparison in actual use to be done.

Additionally, it may well be that the existing file can still provide a good demonstration on how to do development of this code in an opwnwrt environment with quick cycle-times and access to the development codebase (which would be another reason to keep it).  I can see that this is where Docker is used, and adding that to a build often has lead to making a codebase more difficult to quickly develop, debug and fix.

File: CMakeLists.txt
Comment (hamishcoleman) [inline CMakeLists.txt]: The cmake build system for this project is deprecated and should not be used

**Comment by 1715173329** [comment]
> Yes, the CMake files are not well maintained and have not always placed things in the right place.

Where CMake puts things it depends on CMakeLists, but I'm also not familiar with CMake system. These parts can be dropped.

File: packages/openwrt/Makefile
Comment (hamishcoleman) [inline packages/openwrt/Makefile:120]: The almost complete rewrite of this file makes it hard to review.

Additionally, it looks to me like you have copied the new file contents from another repository - a separately maintained fork of a OpenWRT build appears to have a similar file.  Can you document the upstream for this file comes from?  Can you explain any differences from the upstream?  Is there any way we can simply use the upstream file without copying it into this project?

File: packages/openwrt/README.md
Comment (hamishcoleman) [inline packages/openwrt/README.md:48]: This is another file that appears to be basically a rewrite.  Some of the changes here do not appear to improve the documentation, they look like they have deleted some explanatory text.  Is there some reasoning behind these changes that you can expand on?

File: packages/openwrt/etc/init.d/edge
Comment (hamishcoleman) [inline packages/openwrt/etc/init.d/edge]: Changing the name of the binary breaks existing user's expectations.  It also breaks compatibility between this and the non openwrt versions

File: packages/openwrt/etc/init.d/edge
Comment (hamishcoleman) [inline packages/openwrt/etc/init.d/edge:20]: You appear to have removed the stop_service() completely, is that expected?

File: packages/openwrt/etc/init.d/supernode
Comment (hamishcoleman) [inline packages/openwrt/etc/init.d/supernode]: See above, with reference to changing the binary name

File: packages/openwrt/etc/init.d/supernode
Comment (hamishcoleman) [inline packages/openwrt/etc/init.d/supernode:17]: See above, same question about removing stop_service

File: packages/openwrt/Makefile
Comment (hamishcoleman) [inline packages/openwrt/Makefile]: If you are splitting this into small packages for embedded use, It seems that some of these utils can be used by the daemon in regular operations and some are only used for testing or configuration.  So, this may not be the best set to install in each package

**Review by hamishcoleman — CHANGES_REQUESTED** [review]

**Comment by hamishcoleman** [comment]
Have you noticed that there are multiple build failures in the CI?  Is there a known cause for that?

File: .github/workflows/openwrt.yml
Comment (hamishcoleman) [inline .github/workflows/openwrt.yml]: Uploading each different build to a different artifact container makes it hard to see the artifacts.  Have you considered creating one artifact for "packages-ipkg", just like we are doing with the RPM and dpkg builds?

File: packages/openwrt/etc/init.d/edge
Comment (1715173329) [reply packages/openwrt/etc/init.d/edge:20]: Yes, procd will automatically stop the program by sending singal `SIGTERM`.
See https://forum.openwrt.org/t/procd-shutdown-issues-questions/33759/2

**Comment by 1715173329** [comment]
> Have you noticed that there are multiple build failures in the CI? Is there a known cause for that?

Yeah. It's caused by a feature `AUTOREMOVE` from OpenWrt, I'm looking for a way to fix it.

File: packages/openwrt/etc/init.d/edge
Comment (1715173329) [reply packages/openwrt/etc/init.d/edge]: Okay I'll revert this change.

File: .github/openwrt/Dockerfile
Comment (1715173329) [reply .github/openwrt/Dockerfile:6]: https://github.com/ntop/n2n/blob/06c489fd8ad42d6c025beaea0fd62d7d4d948c31/.github/workflows/openwrt.yml#L104-L106

This fixes the runtime tests. The docker image is only used for development purpose so upstream doesn't provide prebuilt one. You can see this is copied from openwrt/packages as-is.

File: .github/openwrt/entrypoint.sh
Comment (1715173329) [reply .github/openwrt/entrypoint.sh:10]: Not yet.

File: .github/openwrt/entrypoint.sh
Comment (1715173329) [reply .github/openwrt/entrypoint.sh:26]: Mentioned above, from github.com/openwrt/packages. `svn` can be used to download these code if you like.

File: .github/openwrt/entrypoint.sh
Comment (1715173329) [reply .github/openwrt/entrypoint.sh:14]: > Can you explain what a "variant" and what this "release" is?

"variant", is just the name defined via Package/`foo`.

"release" is a part of the OpenWrt's version protocol system, see https://openwrt.org/docs/guide-developer/packages

- PKG_VERSION - `The upstream version number that we're downloading`
- PKG_RELEASE - `The version of this package Makefile`

File: .github/openwrt/test.sh
Comment (1715173329) [reply .github/openwrt/test.sh:12]: It's just a rewritten of `make tests` I think. We cannot run that command in openwrt env.

File: .github/openwrt/test.sh
Comment (1715173329) [reply .github/openwrt/test.sh]: > This doesnt look like a very useful test. It would probably make more sense to check the exit code of running the binary.

I have no idea. Do a actual runtime test can be very complex. So just do a version check here instead. If the binrary is broken it will also fail.

File: .github/workflows/openwrt.yml
Comment (1715173329) [reply .github/workflows/openwrt.yml]: > Failing that, it might be simpler if there were two different files - the file you are rewriting is not broken could exist side-by-side with your replacement. This would allow a comparison in actual use to be done.

I'm not sure what you want. You may simply change `Diff view` to `Split`.

Comment (1715173329) [reply .github/workflows/openwrt.yml]: > Additionally, it may well be that the existing file can still provide a good demonstration on how to do development of this code in an opwnwrt environment with quick cycle-times and access to the development codebase (which would be another reason to keep it).

If you want to do some developments, the proper approach is to download and use the SDK provided by upstream. In this case building toolchain from source is time-wasting IMHO.

> I can see that this is where Docker is used, and adding that to a build often has lead to making a codebase more difficult to quickly develop, debug and fix.

Mentioned above, just used for runtime tests. This is crossing-compile, we cannot do that on host.

File: CMakeLists.txt
Comment (1715173329) [reply CMakeLists.txt]: Okay, I will drop this change.

File: .github/workflows/openwrt.yml
Comment (1715173329) [reply .github/workflows/openwrt.yml]: I'm not familiar with `actions/upload-artifact`. Can it be called multiple times with the same name?

File: packages/openwrt/Makefile
Comment (1715173329) [reply packages/openwrt/Makefile:120]: It's from https://github.com/immortalwrt/packages/blob/master/net/n2n/Makefile, which is copied from this repo, and rewritten by me recently, then I submit these changes back.

In ImmortalWrt, we used more user-friendly uci config to generate the final configurations. Also merged `n2n-edge` and `n2n-supernode` into a single `n2n` package. Of course, one is need to download the remote source code and another is built from local source.

My $0.02 is not to rely on other projects as much as we can.

File: packages/openwrt/README.md
Comment (1715173329) [reply packages/openwrt/README.md:48]: Yeah, I just updated the build instructions and some outdated lines were removed. 
"explanatory text" is at OpenWrt docs where I pointed out.

File: packages/openwrt/Makefile
Comment (1715173329) [reply packages/openwrt/Makefile]: I'm not familiar with them. It would be best if you can give me some advice.

**Comment by hamishcoleman** [comment]
Each time you force push, it not only makes it annoying to see your change, but much more importantly, there is no comment or context for /why/ you make whatever change you made.  If you simply add new commits, each one can have a command explaining the commit - before merging, we could easily squash the extra commits that this might cause (or, perhaps better, leave them so there is a track and trail for others who are attempting to understand why the code was written in the way it is)

**Comment by hamishcoleman** [comment]
I suspect that you are using some kind of automated yaml formatting tool - unless the tool can avoid making random whitespace changes between each commit, it makes it harder to see the changes that have happened.  This is especially clear with the dancing between two different types of yaml string blocks depending on the length of the string - which makes the differences touch more lines than needed for the change.

**Comment by 1715173329** [comment]
Uhhhh, different community has different rules. For me I just followed rules from OpenWrt (by default).

**Comment by 1715173329** [comment]
> I suspect that you are using some kind of automated yaml formatting tool - unless the tool can avoid making random whitespace changes between each commit, it makes it harder to see the changes that have happened. This is especially clear with the dancing between two different types of yaml string blocks depending on the length of the string - which makes the differences touch more lines than needed for the change.

Aha, are u sure? All changes were made by hand.

**Comment by hamishcoleman** [comment]
> > [...] This is especially clear with the dancing between two different types of yaml string blocks depending on the length of the string - which makes the differences touch more lines than needed for the change.
> 
> Aha, are u sure? All changes were made by hand.

No, it was a guess - based on other times I have seen this.

Ideally, try to minimise the number of lines your patch touches - so if a string block is already multi-line, just leaving it
as a multi-line allows the patch to touch less lines.

(Obviously if a single-line string needs to be split for line-length reasons, this still should be done)

**Comment by 1715173329** [comment]
> Ideally, try to minimise the number of lines your patch touches - so if a string block is already multi-line, just leaving it
as a multi-line allows the patch to touch less lines.

My $0.02 is to keep the final file as tiny as I can.

> (Obviously if a single-line string needs to be split for line-length reasons, this still should be done)

Yeah, but this line isn't that so long and wrapping a new line doesn't really help (due to indentation.)

**Comment by 1715173329** [comment]
Based on the comments above, the maintainers might not use openwrt for development, so I decide to close this PR, as maintaining something unfamiliar is not a good idea.
feel free ping me if there's anyone really interested in this.

**Pull request closed without merging**
